ctrl_scripts/precache-slurm.pl

```diff
@@ -0,0 +1,314 @@
+  my ($master_id, $task_id) = $job_id =~ /(\d+)_(\d+)/;
+  return unless $master_id && $task_id;
+
+  my $error_file = "$log_dir/slurm-${master_id}_${task_id}.err";
+  my $output_file = "$log_dir/slurm-${master_id}_${task_id}.out";
+  
+  print "\nJob array task $task_id (master job $master_id) error log:";
+  if (-f $error_file) {
+    print "\n=== First 10 lines of stderr ===";
+    system("head -n 10 $error_file");
+  }
+  if (-f $output_file) {
+    print "\n=== First 10 lines of stdout ===";
+    system("head -n 10 $output_file");
+  }
+}
+
+# Increase resource limits for failed jobs
+sub get_adjusted_resources {
+  my ($idx, $state) = @_;
+  my $current = $job_resources{$idx} || {mem => $default_mem, time => $default_time};
+  $state ||= '';
+
+  if ($state eq 'OUT_OF_MEMORY') {
+    my $mem = $current->{mem} * 2;
+    $mem = $mem > $max_mem ? $max_mem : $mem;
+    return {%$current, mem => $mem};
+  }
+
+  if ($state eq 'TIMEOUT') {
+    my $time = $current->{time} * 2;
+    $time = $time > $max_time ? $max_time : $time;
+    return {%$current, time => $time};
+  }
+  return $current;
+}
+
+# Resubmit failed jobs
+sub handle_job_failure {
+  my ($idx, $job_id, $state) = @_;
+  warn "Job $job_id-$idx failed with $state (attempt $retry_count{$idx}/$max_retries)\n";
+
+  if ($state eq 'OUT_OF_MEMORY' || $state eq 'TIMEOUT') {
+    $job_resources{$idx} = get_adjusted_resources($idx, $state);
+  } else {
+    check_job_logs($job_id);
+  }
+
+  if ($retry_count{$idx} >= $max_retries) {
+    delete $job_ids{$idx};
+    delete $job_resources{$idx};
+    save_status();
+    return 0;
+  }
+  delete $job_ids{$idx};
+  return submit_job($idx);
+}
+
+# Submit a job array
+sub submit_job {
+  my ($start_idx, $end_idx) = @_;
+  return 0 unless defined $start_idx;
+  $end_idx ||= $start_idx;
+  
+  my $array_size = $end_idx - $start_idx + 1;
+  my $resources = get_adjusted_resources($start_idx);
+  
+    my $cmd = qq{sbatch --parsable --array=$start_idx-$end_idx }.
+            qq{--output=$log_dir/slurm-%A_%a.out --error=$log_dir/slurm-%A_%a.err }.
+            qq{--time=$resources->{time}:00:00 --mem=$resources->{mem}G }.
+            qq{--wrap='perl $libs $Bin/precache.pl --mode=index --index=\$SLURM_ARRAY_TASK_ID'};
+  
+  $verbose && print "Submitting: $cmd\n";
+
+  chomp(my $array_job_id = qx($cmd));
+  if ($?) {
+    warn "sbatch command failed with status $?";
+    return 0;
+  }
+
+  # Register submitted jobs for tracking
+  for my $idx ($start_idx..$end_idx) {
+    $job_ids{$idx} = "${array_job_id}_$idx";
+    $retry_count{$idx}++;
+  }
+  save_status();
+  return 1;
+}
+
+# Return current state of a job
+sub get_job_state {
+  my ($job_id) = @_;
+  my $cmd = "sacct -j $job_id --format=state -n --parsable2 | head -n1";
+  chomp(my $state = qx($cmd));
+  $state =~ s/\s+//g;
+  return $state || 'UNKNOWN';
+}
+
+print "Submitting $njobs precache jobs in $max_submissions x $max_array_size batches...\n";
+
+# Process & monitor all jobs
+while (1) {
```

Comment:
Is this script ever supposed to exit on its own?

Reply:
pls, see "last" at line 307

Reply:
Yes, the exit condition (all jobs are done) is at the end of the loop (was easier to implement than at the start of the loop).